### DISCUSSION TITLE
Dynaview - Make it work for all scrollables and allow saving the topmost tiddler

### DISCUSSION THREAD

**Discussion by BurningTreeC** [TiddlyWiki/TiddlyWiki5#5825] [Ideas]
Hello there,

I wonder if we can make the dynaview plugin work so that it doesn't only listen on the window for scroll events but on all scrollable elements

Further, a configuration option would be great, which allows saving the topmost tiddler title in a state tiddler if set to yes

What do you think?

**Comment by Jermolene** [comment]
> What do you think?

Great idea, would you be interested in working on that?

Alternatively, we could explore a different approach using intersection observers. I think that would require us to move to a widget-based architecture, where there would be a `<$dynaview>` widget that sets up an intersection observer on its content, and we wouldn't do all the DOM attribute-based communication.

**Reply by saqimtiaz** [reply]
@Jermolene looks like we posted simultaneously.
+1 for leveraging intersection observers.

**Reply by saqimtiaz** [reply]
Two things that might be useful to consider if we work on this:

- with `dynaview` I managed to set it up with my Streams plugin so that it wasn't required but was automatically used if present. This required a fair bit of wrangling with the templates. It would be good if we could keep such use cases in mind  in the widget-based approach and see if we can make them easier to implement.
- one of the issues that I ran into (and this may only have been in my modified version) was that it was only effective for the first display of a tiddler. Once it had been rendered the first time, the associated state tiddler already exists. So if the tiddler was closed and then re-opened somewhere off screen, it would still be rendered.

**Reply by BurningTreeC** [reply]
> > What do you think?
> 
> Great idea, would you be interested in working on that?
> 
> Alternatively, we could explore a different approach using intersection observers. I think that would require us to move to a widget-based architecture, where there would be a `<$dynaview>` widget that sets up an intersection observer on its content, and we wouldn't do all the DOM attribute-based communication.

@Jermolene, yes I'd be interested in working on that

**Comment by saqimtiaz** [comment]
> I wonder if we can make the dynaview plugin work so that it doesn't only listen on the window for scroll events but on all scrollable elements

This is a good idea. I wrote my own version of dynaview (that I call dynascroll) for my Notation editor. The idea was always to eventually modify the original to be more flexible. I've also experimented with using [intersection observer](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API) to implement the same functionality, however there is no IE support for it.

**Reply by BurningTreeC** [reply]
Thanks @saqimtiaz , there's a polyfill for the intersection observer if we want to implement it that way

**Comment by BurningTreeC** [comment]
@Jermolene @saqimtiaz ,

if we create such a `<$dynaview>` widget, how do we listen on the `window` for scroll events?

**Reply by saqimtiaz** [reply]
The great thing about intersection observer is that we don't need to listen to scroll events.

Have a look at the examples of code given here: https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API

**Reply by pmario** [reply]
IMO there is a reason, why "intersection observer" exists. ... Because we avoid to listen on scroll events if possible. This event fires a million times and has the potential to slow down the whole UI and it creates so called "scroll jank". 

Since the browser doesn't know if the event handler stops propagation it has to wait until the handler is finished. If the handler is slow this causes a visual "jank"

**Reply by Jermolene** [reply]
> IMO there is a reason, why "intersection observer" exists. ... Because we avoid to listen on scroll events if possible. This event fires a million times and has the potential to slow down the whole UI and it creates so called "scroll jank". 
> 
> 
> 
> Since the browser doesn't know if the event handler stops propagation it has to wait until the handler is finished. If the handler is slow this causes a visual "jank"

The original Dynaview didn't use intersection observer because it wasn't widely implemented at the time. Using scroll events really isn't so bad; the necessary techniques are well understood and work reliably.

**Comment by BurningTreeC** [comment]
@Jermolene @saqimtiaz @pmario,

would such a `<$dynaview>` widget be placed in the `$:/core/ui/PageTemplate` tiddler (like, globally) or in the `$:/core/ui/ViewTemplate` and `$:/core/ui/EditTemplate` tiddlers, which, for me, would make more sense?

**Reply by Jermolene** [reply]
Hi @BurningTreeC the `<$dynaview>` widget would be wrapped around any individual content items that need to be rendered dynamically. I wouldn't have thought it would be included in any of the core templates by default.

**Reply by BurningTreeC** [reply]
Thanks @Jermolene, yes I understand that it wouldn't be included in any core template by default, I'm just trying to get the logic

**Comment by BurningTreeC** [comment]
@Jermolene , I'm experimenting with a widget that uses the `IntersectionObserver` right now and I think that it can work out

**Reply by BurningTreeC** [reply]
The `IntersectionObserver` approach for me doesn't work out that well at the moment. When I reorder tiddlers in the story river using the `Sidebar Open Tab` the `IntersectionObservers` seem to be gone

**Reply by Jermolene** [reply]
Hi @BurningTreeC that sounds like a problem with refreshing the widget. It may yet be problematic that we don't have a "destroy()" method on widgets for cleaning up the observer when they are destroyed.

**Reply by BurningTreeC** [reply]
Yes @Jermolene, sounds like a refresh problem. But I cannot figure out where the problem is. It's true that such a `destroy()` method would be useful here...

**Reply by BurningTreeC** [reply]
@Jermolene, how can such a `destroy()` method be realized?

**Reply by Jermolene** [reply]
> @Jermolene, how can such a `destroy()` method be realized?

That, sadly, is quite a big job but would likely improve memory usage and performance. The core would need updating to recursively call this new destroy method when removing widget nodes; we'll need some sort of test rig so that we can easily verify that widgets are being destroyed as expected.

Alternatively, we can continue to rely on the JS garbage collection engine, which means that one needs to be very careful in event handlers to ensure that the associated DOM node hasn't already been removed from the main DOM tree.

**Reply by BurningTreeC** [reply]
@Jermolene, how can we detect if a widget node is removed?

**Reply by BurningTreeC** [reply]
Should we use a `MutationObserver`?

**Reply by Jermolene** [reply]
Widgets that generate DOM nodes can test whether those DOM nodes are still part of the document using `$tw.utils.domContains(document,element)`.

**Reply by BurningTreeC** [reply]
The first thing that comes to my mind would be to test in the `refresh()` method for `$tw.utils.domContains(document,element)` ... is that a stupid thing?

**Reply by Jermolene** [reply]
> The first thing that comes to my mind would be to test in the `refresh()` method for `$tw.utils.domContains(document,element)` ... is that a stupid thing?

The refresh method won't get called on widgets that have been removed from the tree. It's in the async event handler that you need to check, in case the widget got removed in the interval before the event fired.

**Reply by BurningTreeC** [reply]
Thank you @Jermolene, where is that async event handler?

**Reply by Jermolene** [reply]
> Thank you @Jermolene, where is that async event handler?

Apologies, I'm talking about the handler function for the events emitted by the intersection observer.

At render time, the dynaview widget would output a DOM node with an intersection observer attached. The handler function needs to check that the DOM node is still in the document before doing its work.

**Reply by BurningTreeC** [reply]
@Jermolene, ah now I understand ... I can test for `$tw.utils.domContains()` and call observer.disconnect() if it isn't in the DOM anymore

**Reply by BurningTreeC** [reply]
@Jermolene, I'm now checking in the handler function if the DOM node is still in the document, before the handler function does its work. But the problem persists

**Reply by Jermolene** [reply]
> @Jermolene, I'm now checking in the handler function if the DOM node is still in the document, before the handler function does its work. But the problem persists

Hi @BurningTreeC are you able to share the code?

**Reply by BurningTreeC** [reply]
@Jermolene, yes I will do so this evening 👍

**Reply by Jermolene** [reply]
> @Jermolene, yes I will do so this evening 👍 

Great, thank you!

**Reply by BurningTreeC** [reply]
@Jermolene , here's my code:

```
/*\
title: $:/plugins/tiddlywiki/dynaview/modules/widgets/dynaview.js
type: application/javascript
module-type: widget

Dynaview widget

\*/
(function(){

/*jslint node: true, browser: true */
/*global $tw: false */
"use strict";

var Widget = require("$:/core/modules/widgets/widget.js").widget;

var DynaviewWidget = function(parseTreeNode,options) {
	this.initialise(parseTreeNode,options);
};

/*
Inherit from the base widget class
*/
DynaviewWidget.prototype = new Widget();

/*
Render this widget into the DOM
*/
DynaviewWidget.prototype.render = function(parent,nextSibling) {
	var self = this;
	// Remember parent
	this.parentDomNode = parent;
	// Compute attributes and execute state
	this.computeAttributes();
	this.execute();

	var domNode = this.document.createElement("div");
	domNode.className = "tc-dynaview";

	// Insert element
	parent.insertBefore(domNode,nextSibling);
	this.renderChildren(domNode,null);
	this.domNodes.push(domNode);
	this.domNode = domNode;

	var dynaviewThresholdArray = [];
	var numSteps = 100;

	for (var i=1.0; i<=numSteps; i++) {
		var ratio = i / numSteps;
		dynaviewThresholdArray.push(ratio);
	}

	dynaviewThresholdArray.push(0);

	var dynaviewRoot = null;
	if(this.dynaviewRootSelector) {
		try {
			dynaviewRoot = this.document.querySelector(this.dynaviewRootSelector);
		} catch(e) {

		}
	}

	var options = {
		root: dynaviewRoot,
		rootMargin: this.dynaviewRootMargin,
		threshold: dynaviewThresholdArray
	}

	this.intersectionObserver = new IntersectionObserver(function(entries,observer) {
		if(!$tw.utils.domContains(self.document,self.domNode)) {
			observer.disconnect();
			return;
		}
		$tw.utils.each(entries,function(entry) {
			$tw.wiki.addTiddler(new $tw.Tiddler({title: "$:/state/DynaView/Visibility/" + self.dynaviewTiddlerTitle, ratio: entry.intersectionRatio}));
			if(entry.intersectionRatio > 0) {
				var tiddlerTitles = self.wiki.filterTiddlers("[all[tiddlers+shadows]prefix[$:/state/DynaView/Visibility/]]");
				var highestRatio = 0.0,
					highestRatioTiddler;
				$tw.utils.each(tiddlerTitles,function(tiddlerTitle) {
					var tiddler = self.wiki.getTiddler(tiddlerTitle);
					if(tiddler && tiddler.fields) {
						var currentTiddler = tiddler.fields.title.replace("$:/state/DynaView/Visibility/","");
						var currentRatio = tiddler.fields.ratio;
						if(currentRatio >= highestRatio) {
							highestRatio = currentRatio;
							highestRatioTiddler = currentTiddler;
						}
					}
				});
				if(highestRatioTiddler) {
					$tw.wiki.addTiddler(new $tw.Tiddler({title: "$:/state/DynaView/topmost-tiddler", text: highestRatioTiddler}));
				} else {

				}
			} else if(entry.intersectionRatio === 0) {
			}
		});
	},options);

	var dynaviewNode;
	if(this.dynaviewNodeSelector) {
		try {
			dynaviewNode = this.document.querySelector(this.dynaviewNodeSelector);
		} catch(e) {

		}
	}

	console.log(dynaviewNode);
	console.log($tw.utils.domContains(this.document,domNode));

	if(dynaviewNode && $tw.utils.domContains(this.document,domNode)) {
		this.intersectionObserver.observe(dynaviewNode);
	}
};

/*
Compute the internal state of the widget
*/
DynaviewWidget.prototype.execute = function() {
	// Get attributes
	this.dynaviewTiddlerTitle = this.getAttribute("tiddler-title");
	this.dynaviewNodeSelector = this.getAttribute("node-selector");
	this.dynaviewRootSelector = this.getAttribute("root",null);
	this.dynaviewThreshold = parseFloat(this.getAttribute("threshold","0"));
	this.dynaviewRootMargin = this.getAttribute("root-margin","0px");
	// Make child widgets
	this.makeChildWidgets();
};

/*
Selectively refreshes the widget if needed. Returns true if the widget or any of its children needed re-rendering
*/
DynaviewWidget.prototype.refresh = function(changedTiddlers) {
	var changedAttributes = this.computeAttributes();
	if(changedAttributes.root || changedAttributes.threshold || changedAttributes["root-margin"] || changedAttributes["node-selector"] || changedAttributes["tiddler-title"]) {
		this.intersectionObserver.disconnect();
		this.refreshSelf();
		return true;
	}
	return this.refreshChildren(changedTiddlers);
};

exports.dynaview = DynaviewWidget;

})();

```

**Reply by BurningTreeC** [reply]
@Jermolene, any thoughts?

**Reply by pmario** [reply]
can you post an example link where it is used  or a zip here, Similar to https://tiddlywiki.com/plugins/tiddlywiki/dynaview/

**Reply by Jermolene** [reply]
Hi @BurningTreeC it would indeed be helpful to see it in action, eg it's not clear what the attributes do.

**Reply by BurningTreeC** [reply]
[index.zip](https://github.com/Jermolene/TiddlyWiki5/files/6759649/index.zip)

@Jermolene @pmario, here's a Demo

**Reply by BurningTreeC** [reply]
You can see that it stops working if you reorder tiddlers in the `SideBar/Open` tab

**Reply by pmario** [reply]
Why does the EditTemplate need a `<$dynaview>` widget? 

My understanding of the dynaview is, that is should speed up the page rendering if many elements are part of the the eg: StoryRiver but are not shown in the browser viewport. 

I don't think there is any performance penalty, if we show +- 10 elements in edit view.

**Reply by Jermolene** [reply]
Hi @BurningTreeC I'm still struggling to understand the demo, and what is supposed to be happening. But the overview of the widget attributes is useful. I wasn't expecting the "node-selector" attribute, I was thinking that the intersection would be observed on the DIV generated by the dynaview widget. What are the use cases for specifying a different "root" attribute?

**Reply by BurningTreeC** [reply]
> Why does the EditTemplate need a `<$dynaview>` widget?
> 
> My understanding of the dynaview is, that is should speed up the page rendering if many elements are part of the the eg: StoryRiver but are not shown in the browser viewport.
> 
> I don't think there is any performance penalty, if we show +- 10 elements in edit view.

@pmario , the goal is to also get the "topmost tiddler" from the dynaview widget and therefor it's also needed in the EditTemplate

**Reply by pmario** [reply]
At the moment the code shows the "top-most" tiddler in the $:/SiteSubtitle tiddler. ... The problem is, that it doesn't show the "top-most" tiddler. It shows the tiddler with the highest "intersection-ratio"... 

So eg: if a tiddler A is 4 cm high and 100% visible it has ratio of 1
If a tiddler B is 2 cm high and 100% visible it also has the ratio of 1. 

So if A is scrolled up and only 3.9cm are visible it has a ratio of <1 ... So the tiddler B which still has ratio 1 will be top most. ... Which isn't true. 

@BurningTreeC ... IMO there is a conceptual problem.

**Reply by BurningTreeC** [reply]
> Hi @BurningTreeC I'm still struggling to understand the demo, and what is supposed to be happening. But the overview of the widget attributes is useful. I wasn't expecting the "node-selector" attribute, I was thinking that the intersection would be observed on the DIV generated by the dynaview widget. What are the use cases for specifying a different "root" attribute?

@Jermolene,

I didn't think too much about on what node the observer should listen. I'll try changing it so that it listens on the generated dom node.

The Demo tries to show the tiddler in the wiki's subtitle which became visible last

The use cases for specifying a different "root" attribute are if the scrollable element isn't the window or the viewport. For example if we use the dynaview widget within your twpub-tools layout, we would specify a selector for the root attribute

**Reply by BurningTreeC** [reply]
> At the moment the code shows the "top-most" tiddler in the $:/SiteSubtitle tiddler. ... The problem is, that it doesn't show the "top-most" tiddler. It shows the tiddler with the highest "intersection-ratio"...
> 
> So eg: if a tiddler A is 4 cm high and 100% visible it has ratio of 1
> If a tiddler B is 2 cm high and 100% visible it also has the ratio of 1.
> 
> So if A is scrolled up and only 3.9cm are visible it has a ratio of <1 ... So the tiddler B which still has ratio 1 will be top most. ... Which isn't true.
> 
> @BurningTreeC ... IMO there is a conceptual problem.

Yes there is a conceptual problem but I thought we would first try to solve the problem why the intersection-observers stop working when we reorder DOM Nodes using the `SideBar/Open` tab for example

**Reply by BurningTreeC** [reply]
After that I thought we could get the logic right

**Reply by saqimtiaz** [reply]
@BurningTreeC it has been a while since I prototyped something similar with IntersectionObserver, but as I recall I had a startup script setting up an IntersectionObserver for the document viewport with a reference saved as `$tw.rootIntersectioObserver`. Then with a widget like `<$dynaview/>`, in it's `render()` method you would create its DOM node (domNode) and then call `$tw.rootIntersectionObserver.observe(domNode)`. So one intersection observer, many widgets being observed.

**Reply by BurningTreeC** [reply]
@saqimtiaz, 

> @BurningTreeC it has been a while since I prototyped something similar with IntersectionObserver, but as I recall I had a startup script setting up an IntersectionObserver for the document viewport with a reference saved as `$tw.rootIntersectioObserver`. Then with a widget like `<$dynaview/>`, in it's `render()` method you would create its DOM node (domNode) and then call `$tw.rootIntersectionObserver.observe(domNode)`. So one intersection observer, many widgets being observed.

That sounds reasonable 👍

**Reply by pmario** [reply]
> After that I thought we could get the logic right

OK. I see. ...

**Reply by BurningTreeC** [reply]
@saqimtiaz, the problem is that the intersection observer should be variable to detect intersections with other roots, too

**Reply by pmario** [reply]
> @saqimtiaz, the problem is that the intersection observer should be variable to detect intersections with other roots, too

Then we will need to create an observer-manager with "register" functionality. EG in the sidebar when we switch tabs, the dom-structure for the "hidden" tabs are destroyed. ... Prior to hiding the elements we will need to ask the manger if an observer is active for this root and call the disconnect() function. When the tab will get visible again, the observer has to be instantiated again. IMO a simple lookup table should be enough.

**Comment by pmario** [comment]
What I had expected for the story river / classic storyview is the following: 

 - If a `tc-tiddler-frame` gets visible on the screen it will be rendered.
 - If a tiddler-frame leaves the viewport it will not be re-rendered anymore, because it can't be seen. 
 - If a tiddler title reaches eg: the upper area of the current browser viewport, it becomes the top-most tiddler
   - "the upper area" can be defined by the user eg: 20%
   - If the title is below the top 20% of the area, that tiddler at the top is the top-most tiddler, even if the title can't be seen anymore. 

That's basically it for the StoryRiver. 

For long lists in the right sidebar it is similar, but it wouldn't be the `tc-tiddler-frame` that will be watched. It will be `tc-menu-list-item`

**Reply by BurningTreeC** [reply]
👍 @pmario
